server/src/main/java/io/spine/server/bus/Bus.java

```diff
@@ -99,4 +136,172 @@ public void unregister(D dispatcher) {
+
+    /**
+     * Packs the given message of type {@code T} into an envelope of type {@code E}.
+     *
+     * @param message the message to pack
+     * @return new envelope with the given message inside
+     */
+    protected abstract E toEnvelope(T message);
+
+    /**
+     * Posts the given envelope to the bus.
+     *
+     * <p>Finds and invokes the {@linkplain MessageDispatcher MessageDispatcher(s)} for the given
+     * message.
+     *
+     * <p>This method assumes that the given message has passed the filtering.
+     *
+     * @see #post(Message, StreamObserver) for the public API
+     */
+    protected abstract void doPost(E envelope, StreamObserver<?> failureObserver);
+
+    /**
+     * Posts each of the given envelopes into the bus and acknowledges the message posting with
+     * the {@code acknowledgement} observer.
+     *
+     * @param envelopes       the envelopes to post
+     * @param acknowledgement the observer of the message posting
+     */
+    private void doPost(Iterable<E> envelopes, StreamObserver<T> acknowledgement) {
+        final CountingStreamSupervisor<T> ackingSupervisor =
+                new CountingStreamSupervisor<>(acknowledgement);
+        int currentErrorCount = ackingSupervisor.getErrorCount();
+        for (E message : envelopes) {
+            doPost(message, ackingSupervisor);
+            if (currentErrorCount == ackingSupervisor.getErrorCount()) {
+                ackingSupervisor.onNext(message.getOuterObject());
+            } else {
+                currentErrorCount = ackingSupervisor.getErrorCount();
+            }
+        }
+        if (ackingSupervisor.getErrorCount() == 0) {
+            ackingSupervisor.onCompleted();
+        }
+    }
+
+    /**
+     * Stores the given messages into the underlying storage.
+     *
+     * @param messages the messages to store
+     */
+    protected abstract void store(Iterable<T> messages);
+
+    /**
+     * @return a {@link Function} converting the messages into the envelopes of the specified
+     * type
+     */
+    protected final Function<T, E> toEnvelope() {
+        return messageConverter;
+    }
+
+    /**
+     * @return a {@link Function} converting the envelopes into the messages of the specified
+     * type
+     */
+    protected final Function<E, T> toMessage() {
+        return messageConverter.reverse();
+    }
+
+    /**
+     * A function creating the instances of {@link MessageEnvelope} from the given message.
+     */
+    private class MessageToEnvelope extends Converter<T, E> {
+
+        @Override
+        public E doForward(T message) {
+            final E result = toEnvelope(message);
+            return result;
+        }
+
+        @Override
+        protected T doBackward(E envelope) {
+            final T result = envelope.getOuterObject();
+            return result;
+        }
+    }
+
+    /**
+     * Supervises the underlying {@link StreamObserver} and performs a defined action on a call to
```

Comment:
This is just a facade for a stream observer. It allows to spy on the values passed. I would call it SpyObserver, add methods to optionally spy on `next` and `completed` events. Then document it properly and move to `StreamObservers`.